server/controllers/finance/trialBalance/index.js

```diff
@@ -1,126 +1,99 @@
-      WHERE posting_journal.trans_id IN (?)
-      GROUP BY posting_journal.account_id
-    ) AS combined
-      JOIN account ON account.id = combined.account_id
-      ORDER BY account.number;
-  `;
-
-  // execute the query
-  db.exec(sql, [transactions, transactions])
-    .then(data => res.status(200).json(data))
-    .catch(next);
-};
-
-/**
- * @function checkTransactions
- * @descriptions
- * fires all check functions and return back an array of promisses containing the errors
- * here are the list of checks [type of error]:
- *
- * 1. A transaction should have at least one line [FATAL]
- * 2. A transaction must be balanced [FATAL]
- * 3. A transaction must contain unlocked account only [FATAL]
- * 4. A transaction must not miss a account [FATAL]
- * 5. A transaction must not miss a period or a fiscal year [FATAL]
- * 6. A transaction must have every date valid  [FATAL]
- * 7. A transaction must have a ID for every line [FATAL]
- * 8. A transaction must have an ID of an entity for every line which have a no null entity type [FATAL]
- * 9. A transaction should have an entity ID if not a warning must be printed, but it is not critical [WARNING]
- *
- *
- * Here is the format of error and warnings :
- *
- * exceptions : [{
- *     code : '',          // e.g 'MISSING_ACCOUNT'
- *     fatal : false,      // true for error (will block posting) and false for warning (will not block posting)
- *     transactions : ['HBB1'],   // affected transaction ids list
- *     affectedRows : 12          // number of affectedRows in the transaction
- *   }]
- **/
-exports.checkTransactions = function runTrialBalance(req, res, next) {
+exports.runTrialBalance = function runTrialBalance(req, res, next) {
   const transactions = req.body.transactions;
 
-  if (!transactions) {
-    return next(new BadRequest('The transaction list is null or undefined'));
-  }
-
-  if (!Array.isArray(transactions)) {
-    return next(new BadRequest('The query is bad formatted'));
+  try {
+    validateTransactions(transactions);
+  } catch (e) {
+    return next(e);
   }
 
   const txn = db.transaction();
 
-  transactions.forEach((transaction) => {
-    txn.addQuery('CALL StageTrialBalanceTransaction(?);', transaction);
-  });
+  // stage all trial balance transactions
+  stageTrialBalanceTransactions(txn, transactions);
 
-  txn.addQuery('CALL TrialBalance()');
+  // compute any trial balance errors
+  txn.addQuery('CALL TrialBalanceErrors()');
+
+  // compute the trial balance summary
+  txn.addQuery('CALL TrialBalanceSummary()');
 
   return txn.execute()
-    .then((errors) => {
-      const lastIndex = errors.length - 1;
-      res.status(201).json(errors[lastIndex][0]);
+    .then((results) => {
+      // because we do not know the number of stageTrialBalance() calls, we must index back by two
+      // to get to the CALL TrialBalanceErrors() query and one for the Call TrialBalanceSummary()
+      // query.
+      const errorsIndex = results.length - 2;
```

Comment:
Depending on what you need to do with the array after this if the arrays you need are guaranteed at the end this could just use `pop`, for example: 
```js
// remove the last two elements, summary and errors queries 
const data = { 
  summary : results.pop()
  errors : results.pop()
};
```

Having read this I'm not even sure it's clearer - just a thought!